<a href="https://colab.research.google.com/github/Emanuel-nx/Estudo_pratico_com_Colab/blob/main/Pr%C3%A1tica_Colab_I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [54]:
from sklearn import datasets
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from xgboost import XGBRegressor
import pandas as pd
import numpy as np

In [63]:
from IPython import get_ipython
from IPython.display import display
# %%
from sklearn import datasets
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from xgboost import XGBRegressor
# Import the sklearn wrapper for compatibility
from xgboost.sklearn import XGBRegressor as XGBRegressor_sklearn
import pandas as pd
import numpy as np


In [2]:
housing = datasets.fetch_california_housing()

In [3]:
df = pd.DataFrame(housing.data, columns=housing.feature_names)

In [4]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [5]:
# Listar o que temos neste dicionario
housing.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'feature_names', 'DESCR'])

In [6]:
print(housing.DESCR)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

:Number of Instances: 20640

:Number of Attributes: 8 numeric, predictive attributes and the target

:Attribute Information:
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

:Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per ce

Conjunto de dados de habitação da Califórnia
--------------------------

**Características do conjunto de dados:**

:Número de instâncias: 20640

:Número de atributos: 8 atributos numéricos, preditivos e o alvo

:Informações de atributo:
- Renda média MedInc no grupo de blocos
- Idade da casa: idade média da casa no grupo de quarteirões
- Número médio de quartos por família AveRooms
- Número médio de quartos por domicílio
- População do grupo de blocos populacionais
- Número médio de membros da família AveOccup
- Latitude do grupo de blocos de latitude
- Longitude do grupo de blocos de longitude

:Valores de atributo ausentes: Nenhum

Este conjunto de dados foi obtido do repositório StatLib.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

A variável-alvo é o valor médio da casa para os distritos da Califórnia,
expresso em centenas de milhares de dólares (US$ 100.000).

Este conjunto de dados foi derivado do censo dos EUA de 1990, usando uma linha por censo
grupo de blocos. Um grupo de blocos é a menor unidade geográfica para a qual os EUA
Escritório do Censo público

In [7]:
print(housing.target)

[4.526 3.585 3.521 ... 0.923 0.847 0.894]


In [13]:
# CARACTERISTICAS DAS CASAS
X = housing.data
# Preços
y = housing.target

# MODELING TECHNIQUES



1.   Regressão Linear do SKLEARN: [sklearn.linear_model.LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression)
2.   Support Vector Regression do SKLEARN: [sklearn.svm.SVR](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html#sklearn.svm.SVR)
3. Decision Tree Regression do XGBoost: [Python API Reference](https://xgboost.readthedocs.io/en/latest/python/python_api.html)


Modeling Assumptions:

Apenas variáveis numéricas  

#TEST DEDIGN

##Dataset split:

Separação de Train/Test dataset padrão com 20% de massa para teste via metodo SKLEARN

In [14]:
X_train, X_test, y_Train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

In [15]:
print(X_train)

[[   3.2596       33.            5.0176565  ...    3.6918138
    32.71       -117.03      ]
 [   3.8125       49.            4.47354497 ...    1.73809524
    33.77       -118.16      ]
 [   4.1563        4.            5.64583333 ...    2.72321429
    34.66       -120.48      ]
 ...
 [   2.9344       36.            3.98671727 ...    3.33206831
    34.03       -118.38      ]
 [   5.7192       15.            6.39534884 ...    3.17889088
    37.58       -121.96      ]
 [   2.5755       52.            3.40257649 ...    2.10869565
    37.77       -122.42      ]]


In [16]:
print(y_Train)

[1.03  3.821 1.726 ... 2.221 2.835 3.25 ]


In [20]:
regLinear = LinearRegression().fit(X_train, y_Train)

In [21]:
yLinear = regLinear.predict(X_test)

In [22]:
MSE = mean_squared_error(y_test, yLinear)

In [23]:
print(MSE)

0.5558915986952422


In [26]:
print("RMSE: ", np.sqrt(MSE))

RMSE:  0.7455813830127749


##Tecnica 02. SVR

In [29]:
regSVR = SVR().fit(X_train, y_Train)

In [30]:
ySVR = regSVR.predict(X_test)

In [31]:
MSESVR = mean_squared_error(y_test, ySVR)

In [32]:
print("MSE SVR: ", MSESVR)
print("RMSE SVR: ", np.sqrt(MSESVR))

MSE SVR:  1.3320115421348744
RMSE SVR:  1.1541280440812771


##Técnica 03. Decision Tree Regression (XGBoost)

In [34]:
regXGB = XGBRegressor().fit(X_train, y_Train)

In [35]:
yXGB = regXGB.predict(X_test)

In [36]:
MSEXGB = mean_squared_error(y_test, yXGB)

In [37]:
print("MSE XGB: ", MSEXGB)
print("RMSE XGB: ", np.sqrt(MSEXGB))

MSE XGB:  0.2225899267544737
RMSE XGB:  0.4717943691423984


##Otimização de Hiperparametros

Utilizando o metodo GridSerachCV do SKLEARN

In [39]:
regXGB.get_params().keys()

dict_keys(['objective', 'base_score', 'booster', 'callbacks', 'colsample_bylevel', 'colsample_bynode', 'colsample_bytree', 'device', 'early_stopping_rounds', 'enable_categorical', 'eval_metric', 'feature_types', 'gamma', 'grow_policy', 'importance_type', 'interaction_constraints', 'learning_rate', 'max_bin', 'max_cat_threshold', 'max_cat_to_onehot', 'max_delta_step', 'max_depth', 'max_leaves', 'min_child_weight', 'missing', 'monotone_constraints', 'multi_strategy', 'n_estimators', 'n_jobs', 'num_parallel_tree', 'random_state', 'reg_alpha', 'reg_lambda', 'sampling_method', 'scale_pos_weight', 'subsample', 'tree_method', 'validate_parameters', 'verbosity'])

In [72]:
Parameters = {
    'max_depth': [5, 6, 7],
    'learning_rate': [0.1, 0.2, 0.3],
    'objective': ['reg: squarederror'],
    'booster': ['gbtree'],
    'n_jobs': [5],
    'gamma': [0, 1],
    'min_child_weight': [1,3],
    'max_delta_step': [0,1],
    'subsample': [0.5, 1]
}